### ISSUE TITLE
CS for DerivedProjectedCRS

### ISSUE THREAD

**Issue by RogerLott** [ISO-TC211/ISO19111#32]
A Derived Projected CRS could be a grid. The DerivedProjectedCRS class therefore needs an association to all three of affineCS, CartesianCS and OrdinalCS.
 
If Cartesian CS remains a subtype of Affine CS the new union would be of AffineCS and OrdinalCS.

But I feel it might be better to not subtype CartesianCS off Affine CS and for the new union class to contain AffineCS, CartesianCS and OrdinalCS. Then CartesianCS (association from ProjectedCRS) would be subtyped directly off CS, and DerivedProjectedCRS associated to the new union class.

I suggest that this new union class be called DerivedProjectedCS (similar to EngineeringCRS associated to EngineeringCS and GeodeticCRS associated to GeodeticCS). But if a more suitable name is available that is fine.

**Comment by jetgeo** [comment]
That would be something like this, except the change for AffineCS/CartesianCS. Awaiting discussion for that one. 

![cs crs associations](https://user-images.githubusercontent.com/7501876/30524602-57e1c980-9bf7-11e7-949d-9d9ae7e1a7f5.png)

**Comment by RogerLott** [comment]
A projectedCRS must have a CartesianCS. It cannot have an affineCS. Seems to me that the modular advantage that Martin argued to be the benefit of subtyping Cartesian off affine disappears once DerivedProjectedCS is introduced. I await Martin's perspective on my Cartesian/Affine relationship comment, but to avoid misunderstanding [at least by neophytes like me] I think the affineCS class should disappear from this diagram and CartesianCS by subtyped directly off CS (which would therefore happen in the CS diagram too). 

One correction needed in the EngineeringCS class, should be ordinalCS:OrdinalCS.

And should the temporalCS attribute be suppressed in this diagram?

**Comment by desruisseaux** [comment]
Maybe I need a little clarification about what is a grid of a derived projected CRS? For example what would be its unit of measurement? If it still a linear unit (meter or related), then I think it would be a different kind of grid than the one for which `OrdinalCS` was introduced.

**Comment by jetgeo** [comment]
OK, see below. 

![cs crs associations](https://user-images.githubusercontent.com/7501876/30531072-1a8aeab2-9c4c-11e7-8182-83462e71f30d.png)

**Comment by RogerLott** [comment]
@desruisseaux asks for clarification about the need for a derived projected CRS using an ordinalCS. Look at example E.2.8 in the 19111 draft. This was originally example 1 in clause 15.5.2 of 19162, CRS WKT, but now onsidered to be incorrect because it inherits a geodetic datum so cannot be a derived engineering CRS. In 19111 example E.2.8 I believe that the CS is not affine and should be ordinal. The bins are just sequential numbers. So in the example we should have (in WKT):
CS[Ordinal,2],
AXIS["(I)",northNorthWest],
AXIS["(J)",westSouthWest],
PSEUDOUNIT["No_unit"] (or PSEUDOUNIT["DiscreteQuantity"], see issue #24).

However here we are in the realms of grids and perhaps this is out of scope.

**Comment by RogerLott** [comment]
Assuming a derivedProjectedCS may be an ordinalCS, the above diagram subtyping CartesianCS directly off CS looks fine to me. Correction for EngineeringCRS.ordinalCS noted.

**Comment by desruisseaux** [comment]
I suggest to keep `CartesianCS` as a sub-type of `AffineCS` if we agree that this relationship is conceptually correct, even if DerivedProjectedCRS accepts also an `OrdinalCS`. I think there is value in capturing the concept that a Cartesian coordinate system is a special case of affine. It allows users to write their own classes (outside ISO 19111) where a CS can be affine or Cartesian but not ordinal.

On the question about allowing `DerivedProjectedCRS` to allows an `OrdinalCS`, taking the example of P6 Seismic Bin Grid Transformation, section 2.3.2.4 of EPSG guidance note currently defines it as a special case of Affine Geometric Transformation. Of course I'm a total newbies in this domain, but are interpolations allowed in a Seismic Bin Grid? Or said otherwise, are fractional coordinate values commonly used on such grid? If the answer is yes, wouldn't it be a case for an `AffineCS` more than an `OrdinalCS`?

**Comment by RogerLott** [comment]
Although currently they are in the definitions of 19111 clause 4 independently, you could define a Cartesian CS as an affine CS in which the axes are all orthogonal and all axes have the same units. So it is correct that cartesianCS could be considered (modelled as) a subtype of  affineCS. This would appear on the CS diagram. Then on the CS-CRS associations diagram a derivedProjectedCS would be a union of affineCS and ordinalCS. You now have to refer to two diagrams to get the whole picture. Does not that make it more difficult to see that a DerivedProjectedCRS can be associated with a cartesianCS? I feel that the diagram above is clear.

However perhaps we need to think more broadly here. I can envisage a DerivedProjectedCRS having a polarCS - EN converted to r(theta). And ENh might be converted to spherical or cylindrical coordinates. So perhaps DerivedProjectedCRS should be associated with EngineeringCS, in which case the DerivedProjectedCS class would not be needed. 

Re your question about seismic bin grid 'coordinates', in my experience they are always integer. The interval may not be one; multiples e.g. 4 on one axis and 1 on the other can be encountered. But I have not seen fractions.

**Comment by desruisseaux** [comment]
Thanks for the information about seismic bin grid. It indeed looks like `OrdinalCS`. Regarding the union of `CartesianCS`, `AffineCS` and `OrdinalCS`, it is indeed easier to see the 3 on a single diagram. But I wonder if we can put `CartesianCS` in the diagram even if it is a subtype of `AffineCS`? That would be redundant, but not a contradiction; I would not see a harm. I believe there is value in reflecting the semantic of coordinate systems in the type hierarchy even if it is not of immediate use to the UML published in ISO 191111.

Regarding `DerivedProjectedCRS` accepting also `PolarCS`, `CylindricalCS` and `SphericalCS`, it looks nice to me. But we have a type issue: `DerivedProjectedCRS` is a subtype of `ProjectedCRS`, and `ProjectedCRS` accepts only coordinate systems of type `CartesianCS`. We can restrict the list of coordinate systems in CRS subtypes, but not expand it. So strictly speaking, in my understanding of UML `DerivedProjectedCRS` as it stands today would not be allowed to be associated to any other CS type than `CartesianCS`. In order to resolve this issue, what about the following approach?

* Rename `DerivedProjectedCRS` into something without "projected" word in it, since I think many users associate "projected" to "map projection" where the target CS is flat. Even if mathematically a projection can be more general, we may be at risk of confusing users if we allow `DerivedProjectedCRS` to be associated to non-flat coordinate systems.
* In the (to be renamed) `DerivedProjectedCRS` class, remove the `ProjectedCRS` parent. Make it extends the `DerivedCRS` type only. This would unlock the use of any coordinate system types.
* Restrict the coordinate system association to `EngineeringCS` as proposed by Roger above.
* Wouldn't the same extended list of CS be relevant to `DerivedGeodeticCRS` too? If the answer is yes, then we could remove `DerivedGeodeticCRS` and `DerivedGeographicCRS` and use the (to be renamed) `DerivedSomethingCRS` class for all of them instead.

I will try to create a UML showing this proposal later.

**Comment by desruisseaux** [comment]
Below is a proposed UML (incomplete because I have no UML tools). It is slightly different than my previous comment. The white boxes are mostly unchanged compared to current draft. The yellow boxes are the core of proposed change. The red arrows and red texts are different compared to current draft.

Basically I swap the position of `ProjectedCRS` and (to be renamed) `DerivedProjectedCRS` so that the parent class accepts wider range of coordinate system types than the children class. Then I tried to find a more generic name for `DerivedProjectedCRS`; I put `DerivedGeodeticCRS` for now but this name is maybe inappropriate in this context. The idea is to have `ProjectedCRS` considered as a special kind of (may need a better name) `DerivedGeodeticCRS`. The same `DerivedGeodeticCRS` class could also be used for deriving CRS from `GeodeticCRS`, if we accept that derived geodetic CRS can have polar CS, cylindrical CS, _etc._ like derived Projected CRS; I do not know if this is correct. So a side-effect of this proposal would be a slight reduction of the amount of classes.

![derivedprojectedcrs](https://user-images.githubusercontent.com/1324414/30689233-a3f162b0-9ec0-11e7-9980-48a682acea99.png)

Note: in my understanding, strictly speaking we would have to include `EllipsoidalCS` in the `GeodeticCS` union because `GeographicCRS` restrict the coordinate system to `EllipsoidalCS`. However I understand the intend to said that `GeodeticCRS` should be associated to `EllipsoidalCS` only if it is actually a `GeographicCRS` subtype. While I'm not sure that omitting `EllipsoidalCS` in `GeodeticCS` is strictly conform to UML, I think that readers will understand the intend.

**Comment by jetgeo** [comment]
I think @desruisseaux is right - "We can restrict the list of coordinate systems in CRS subtypes, but not expand it". I didn't see this conflict until now. This has consequences for Geodetic/Geographic CRS and for Projected/DerivedProjected CRS, as Martin points out. 
For Geodetic/Geographic, we can do as Martin suggest, possibly supplied with a constraint on GeodeticCRS for specifying that it shall not use EllipsoidalCS. Or we can remove the subtyping of Geodetic to Geographic, and have the associations as they are now. 
For Projected/DerivedProjected: The name "DervivedGeodeticCRS" is allready in use. I think this part of the model needs more discussion, and leave that mainly to you geodesy experts.

**Comment by desruisseaux** [comment]
Indeed the `DervivedGeodeticCRS` is already in use, but one of the proposal that I posted uses the same `DervivedGeodeticCRS` for both the existing purpose and for the new derived projected CRS. We don't have to do this way, but it is one of the two proposed ways.

**Comment by jetgeo** [comment]
True, but as I think you have said in the comment three steps above here, using `DerivedGeodeticCRS` for both must include changes  in the inheritance or associations to CS-classes. The existing `DerivedGeodeticCRS` is a subtype of `GeodeticCRS` (and `DerivedCRS`), and can therefore only have `CartesianCS` or `SphericalCS`, while the existing `DerivedProjectedCRS` is supposed to have `AffineCS`, `CartesianCS` or `OrdinalCS`.

**Comment by desruisseaux** [comment]
Yes. My thinking followed those steps:

1. The first proposed fix is a change in class hierarchy for allowing (to be renamed) `DerivedProjectedCRS` to accept a wider range of CS than `ProjectedCRS`.
2. Then I was wondering if `DerivedGeodeticCRS` should also accept a wider range of CS, for similar reasons than `DerivedProjectedCRS`. If and only if the answer is "yes", then:
    1. (to be renamed) `DerivedGeodeticCRS` and `GeodeticCRS` would have the same change in hierarchy than (to be renamed) `DerivedProjectedCRS` and `DerivedCRS`.
    2. As a consequence of above changes, (to be renamed) `DerivedGeodeticCRS` and (to be renamed) `DerivedProjectedCRS` would be at the same level in the class hierarchy, so the same class could be used for both. I picked up `DerivedGeodeticCRS` in the meantime because it seems the most generic one, but another name may be preferable.

**Comment by RogerLott** [comment]
The reason we introduced the geographicCRS subtype is to enforce the requirement for an ellipsoid when the CS is ellipsoidal (because it had been made optional for geodeticCRS with a spherical or [more usually] CartesianCS). If the geodeticCS class must contain ellipsoidal CS, is this effectively sayng that the geographicCRS class does not need to exist and the constraint on GeodeticReferenceframe needs to be suitably amended? Or remove the constrain on GeodeticReferenceframe  and replace it with a constraint on GeodeticCRS to say that if CS=elliposoidal then count(Datum.Ellipsoid)=1 ? 

The model is require to show that:
1) in general
(i) any subtype of SingleCRS may be used as the baseCRS for a derivedCRS
(ii) every subtype of derivedCRS inherits its datum from its baseCRS
(iii) every subtype of derived CRS is constrained to use the same subtype of CS as is its base CRS
    (a) of particular relevance to this discussion is that the allowable subtyes of CS that can be used by a geodeticCRS are cartesian, spherical and ellipsoidal. 
2) by exception
(i) an extra subtype of derivedCRS (called projectedCRS) has been identified because of its practical usage in GIS
(ii) projectedCRS is constrained to use only a CartesianCS;
(iii) projectedCRS can be used as a baseCRS of another derivedCRS [currently called derivedProjectedCRS]; the CRS derived from a projectedCRS (DerivedProjectedCRS) is not constrained to have the same CS as its base projectedCRS. 
   (a) It can have one of several types of CS, notably affineCS, cartesianCS, cylindricalCS, polarCS, sphericalCS and I suppose also userDefinedCS. (VerticalCS, temporalCS, parametricCS and ellipsoidalCS make no sense that I can conceive of inthis context, I am not sure about linearCS).

So if the model were to be as @desruisseaux suggested, 1(iii)a causes problems because it is narrower than 2(iii)a. 
Fixing this by allowing a geodeticCRS to use an engineeringCS or by extending engineeringCS union to include ellipsoidalCS hence allowing an engineeringCRS to use an ellipsoidalCS seem to me to be unacceptable solutions.

if the model were to be as @desruisseaux suggested, to fix the problem with class names what about changing DerivedCRS to GeneralDerivedCRS and then using DerivedCRS for the to-be-changed DerivedGeodeticCRS class.

**Comment by RogerLott** [comment]
Forgot to say in previous post that the condition for a derivedGeodeticCRS being a projectedCRS (as opposed to being a general derivedGeodeticCRS) is that the defining conversion is a map projection.

**Comment by jetgeo** [comment]
> remove the constrain on GeodeticReferenceframe and replace it with a constraint on GeodeticCRS to say that if CS=elliposoidal then count(Datum.Ellipsoid)=1 

Yes, I think that can be a good solution for Geodetic/Geographic.

For Projected/DerivedProjected, a solution may be to use the DerivedProjectedCS union for ProjectedCRS, and add a constraint saying that if type<>DerivedProjected then CS=Cartesian.

I am not sure about the OCL wording, but something like this might work:
![cs crs associations](https://user-images.githubusercontent.com/7501876/30842850-7d7d6638-a285-11e7-84a9-eee810b1f7b3.png)

**Comment by RogerLott** [comment]
This is looking ok for me. But one question. Do we need the projectedCS class? What would be the effect of changing the association from projectedCRS to go to EngineeringCS? We still have the constraint that a projectedCRS can only have a Cartesian CS and adding a few more possible CS types (especially polar, but also cylindrical and spherical) seems appropriate for derivedProjected. But would linear be appropriate for derivedProjectedCRS (as 1D when the CRS is 2D or 3D)?

**Comment by RogerLott** [comment]
Just checking one other issue. Did we agree to remove the UserDefinedCS class? I recollect that At one stage we discussed ordinal CS replacing it. We knew of no use cases. But equally we could not be sure that it had never been used. Mark's meeting notes do not explicitly mention it "Major discussion about implementation of an EngineeringCRS and OrdinalCS; closely linked to underlying implementation of temporal. Significant changes to the model and definitions in re-editing". 
Am I correct in thinking that we concluded that user-definedCS would be deleted from the model and anyone wanting it would be required to create a model extension?

**Comment by desruisseaux** [comment]
Yes, the above model would work. We could also keep `DerivedGeodeticCRS` as it currently stand; part of my proposal had an open question about whether `DerivedGeodeticCRS` should accept a wider range of CS, but I fully agree with Roger's objection against that.

`DerivedProjectedCRS` is a special case because `ProjectedCRS` is already a derived CRS. So `DerivedProjectedCRS` is a derived of a derived CRS, which may need to be handled differently than other cases. An alternative proposal would be to keep almost everything like current draft, except for the following changes:

* Define `DerivedProjectedCRS` as a subtype of DerivedCRS only, not a subtype of `ProjectedCRS`. This solve 2 problems:
    * Avoid conflict with `ProjectedCRS` coordinate systems which was restricted to `CartesianCS`.
    * Avoid conflict with `baseCRS` association, which was toward `GeodeticCRS` in `ProjectedCRS` but was changed toward `ProjectedCRS` in `DerivedProjectedCRS`.
* Add `EllipsoidalCS` in `GeodeticCS` union, as suggested above. Strictly speaking, this is necessary for UML consistency.
* Keep `GeographicCRS` as today. Technically we could omit it, but it seems to me that the wide use of `GeographicCRS` makes it worth to keep.
* Add a constraint in `GeodeticCRS` saying "if the CS is `EllipsoidalCS`, then the CRS type must be `GeographicCRS`".
* Keep `DerivedGeodeticCRS` as it stands today. It can be associated to `CartesianCS`, `SphericalCS` or `EllipsoidalCS`.
* Maybe we don't really need `DerivedGeodeticCRS`? Technically speaking this class is not necessary. We could keep it for symmetry with `GeodeticCRS` but I don't think it is as widely used?

In summary, the core of above proposal is just to change the parent of `DerivedProjectedCRS`.

**Comment by desruisseaux** [comment]
I think we should remove `UserDefinedCS`. Indeed we are not sure that no-one use it. But we already broke compatibility be renaming `GeodeticDatum` as `ReferenceFrame` while I'm pretty sure that `GeodeticDatum` is much more used than `UserDefinedCS`. Another reason is that I think that `UserDefinedCS` type should not have existed in the first place. If a user wants its own CS, (s)he should create his/her own `CoordinateSystem` subtype.

**Comment by RogerLott** [comment]
UserDefined is already removed from the UML but I still have it in the 19111 draft text. Was just checking it was correct to remove it. I am happy to do so.

**Comment by RogerLott** [comment]
@desruisseaux asks whether the DerivedGeodeticCRS class is needed. If you change a CRSs CS, say metres to feet or Cartesian to polar, you can do that in one of two ways (i) create an entirely new CRS, and (optionally) link the two CRSs by a conversion; (ii) create a derivedCRS, which puts the conversion in the derivedCRS definition. Such a unit change applies just as much to a geodeticCRS as the other CRS types: e.g. degrees to radians. geographicCRS could be described as being derived from a geodeticCRS. So yes DerivedGeodeticCRS class is required in the same way that DerivedVerticalCRS or DerivedParametric]CRS might be. 

Just to clarify @desruisseaux proposal: are you suggesting something like this (red optional):
![derivedproj](https://user-images.githubusercontent.com/27864408/30857075-5af8e8bc-a2b2-11e7-960c-f5dde6a8aa84.png)

**Comment by desruisseaux** [comment]
Sorry, my last bullet point was erroneous. It should be read as below (sorry for the error):

* Maybe we don't really need `DerivedGeographcCRS`? Technically speaking this class is not necessary. We could keep it for symmetry with `GeographicCRS` but I don't think it is as widely used?

**Comment by desruisseaux** [comment]
Yes, the schema above captures well my proposal: everything mostly identical to current draft (ignoring the `EllipsoidalCS` in `GeodeticCRS` topic for now), with only `DerivedProjectedCRS` not a `ProjectedCRS` subtype anymore. It is at odd compared to all other `DerivedCRS` subtypes, but `ProjectedCRS` is already a special case.

**Comment by RogerLott** [comment]
I believe that a derivedGeographicCRS is used. See ISO 19162 example 15.3.2 (the CRS name is clearly a copy and paste error which needs to be fixed in the next version!).

@jetgeo please review the diagram in my post above and if the subtyping and associations around ProjectedCRS and DerivedProjectedCRS are acceptable UML then please update the model to conform.
Per my post further above I feel that ProjectedCS should be extended to include +polarCS, +cylindricalCS and +sphericalCS. Is this sufficiently different from EngineeringCS to warrant an additional class? Please decide. Whichever, the constraint on ProjectedCRS should be retained.
Regarding the retention or removal of GeographicCRS and DerivedGeographicCRS, I can live with either. If one is much clearer from a UML perspective then use that. If it is a close decision, then retain these classes, principally on the grounds that the group has not discussed removing them, but I agree with @desruisseaux 's view that "Technically we could omit it, but it seems to me that the wide use of GeographicCRS makes it worth to keep".

**Comment by jetgeo** [comment]
I believe this is done now, according to PPT from @RogerLott  in comment under #19?
See new figures in https://github.com/ISO-TC211/ISO19111/tree/master/Figures

**Comment by desruisseaux** [comment]
Thanks for the new figure. On a related question, in "Coordinate Systems.png", the `AxisDirection` code list has an `unspecified` item. Since code lists are extensible, shouldn't we omit this value and encourage users to define their own value instead? It is not clear to me what would be the difference between allowing an `unspecified` value and making the `axisDirection` attribute optional.
